### DISCUSSION TITLE
Embed in a single-page application

### DISCUSSION THREAD

**Discussion by eino** [Kitware/trame#47] [Q&A] [answered]
Hi,

I have a need for visualization of quite big VTP files, and Trame could be an interesting solutino. However I was wondering about embeddability, as I would need to embed it in a single-page (Vue) application. I see as is, it's a standalone app.

Did someone already embed it in an existin app? Or do you have some ideas of leads / how much work this would imply?

Thanks in advance!

**Comment by jourdain** [comment, marked as answer]
A straight forward way could be to use an iFrame.

If that is not a solution, the core app of trame is vue base but involve some dynamic store/computed prop which might make the merging with yours quite tricky.

Another approach is to include the networking setup within your app and just create the components directly on the JS side while keeping the server "as-is". 

The path forward would depend on the features and integrations that you are looking for. If need be we can help with that work.

**Reply by eino** [reply]
Thanks for your answer. That's what I was worried about. Well, I'll consider the simpler iframe for now, and move to a deeper / more complex integration if it doesn't work out.

**Reply by jourdain** [reply]
One thing that we can make it happen is to enable communication between your app and that iframe so you could technically drive all the app from your UI using [window.postMessage()](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) along with some Python binding on the trame side.

**Reply by eino** [reply]
Actually, I was thinking of your previous message and exploring the problem some more, and it occured that as we have a lot of customization to do, it will probably be easier to start from the separate components rather than the whole Trame.

So, as on client-side we would use vtk.js (in remote rendering mode), I'm not sure about server-side. I see that both Paraview and VTK could be used to provide remote rendering capabilities? What would be the difference? Would VTK work more in an API way, but needing more glue code?

Do you have perchance a documentation (or sample) about getting started with remote rendering?

**Reply by jourdain** [reply]
In term of documentation we have the following set of examples:
- [vtk.js/RemoteView with VTK and/or ParaView](https://kitware.github.io/vtk-js/examples/RemoteView.html)
- [Vue application settings things up with custom RPC](https://github.com/Kitware/web-project-templates/tree/master/vue-vtkjs-pvw-template)
- (Bonus) [trame example doing the same thing as that vue application (1 file / 90 lines)](https://github.com/Kitware/trame/blob/master/examples/VTK/SimpleCone/RemoteRendering.py)

Regarding the different between VTK or ParaView it is really up to you. The code in VTK is more verbose but in exchange you have more controls. ParaView is heavier than plain VTK due to its parallel capabilities. In ParaView you can use the trace to guide you how to do things in Python from GUI interaction. 

HTH,

Seb

**Reply by jourdain** [reply]
Another path could be for you to serve and use your Vue components with-in trame.
